### DISCUSSION TITLE
UI Updates / Custom Events for Select Components

### DISCUSSION THREAD

**Discussion by ldommer** [tasteui/tasteui#356] [Q&A]
### Suggestion Theme

Hi AJ,

congrats to your awesome library :-). I'm using it a lot in one of my projects, after a colleague recommended it to me.

But I'm facing an issue which I'm pretty sure is no bug, but rather a missing feature or me being too blind to figure it out:

I've built a complex filter form for an index table containing multiple select fields. This form also contains a clear all button to reset the state of the form / all fields at once. Simply by calling `$this->reset()` / `$this->resetExcept()` in my Livewire component.

If I click this reset button, the UIs of the select components aren't refreshed. I'm wondering why, because they use `@entangle` under the hood and they are in sync with their Livewire properties, apart from my reset.

Also, if I pass any custom Alpine events to the select components, they won't get passed to the rendered UI and won't be executed.

Are both issues (UI updates and custom events) missing features or is there a way to achieve the wanted behaviour, but I'm to blind to figure it out?

Best regards from Germany

Lennart

### Main Benefits

Better customization and more flexibility.

**Comment by devajmeireles** [comment]
Hey! @ldommer . Welcome to TallStackUI! I hope your experience is amazing. Since this is not an issue, I've converted this to a Discussion.

I don't know if this is your case, but many people seem to not have in-depth knowledge of Livewire to understand how it works so they open issues that are not issues, but in any case, I always respond well to all questions. Feel free to send us anything you want because you think it's important.

1. I created an example of the styled select to reset the properties and I was successful, without any problems. It's worth remembering how `wire:model` works. If you don't make the component hydrate, then the reset will require the bind to occur with `.live`. If you hydrate the component, for example, by calling reset through PHP, then `wire:model` is sufficient to receive the hydration result and nullify the values. This is the example I created:

```php
<?php

namespace App\Livewire;

use Illuminate\Contracts\View\View;
use Livewire\Component;

class Livewire extends Component
{
    public ?int $option = null;

    public function render(): View
    {
        return view('livewire.livewire');
    }

    public function testingReset(): void
    {
        $this->reset();
    }

    public function forceSync(): void
    {
        //
    }
}
```

```blade
<div>
    <x-card>
        @json($option)

        <x-select.styled label="Select One Option"
                            hint="You can choose 1, 2 or 3"
                            :options="[
                                ['label' => 'TALL', 'value' => 1],
                                ['label' => 'LIVT', 'value' => 2],
                            ]"
                            select="label:label|value:value"
                            wire:model="option" />

        <x-button xs wire:click="testingReset">
            Reset
        </x-button>

        <x-button xs wire:click="forceSync">
            Sync
        </x-button>
    </x-card>
</div>
```

2. What are the "Alpine events" that you are passing to the select? You may not need to pass the events, [in favor of using the JS events we provide.](https://tallstackui.com/docs/form/select#styled-events)

**Reply by ldommer** [reply]
Hi AJ, thanks for your long answer and your explanation. I think I was too unprecise (I'm a non-native speaker).

I don't have a problem with wire:model or the core Livewire mechanics or my understanding of these.

The point I tried to describe is the following:

The state of my index component gets properly set back, the table rerenders and all filters are reset. (As all filter fields are wire:model.live). But the UI of the select component does not update. The option I've chosen before keeps staying inside the select box, even if all state is rest. It's only reset if I hit the small x button.

So I tried to pass an Alpine change event, in order to call your select component clear method manually. So in simple words, the select component (UI) does not reset / clear if I call reset on my Livewire component.

**Reply by devajmeireles** [reply]
First of all, you're talking about `select.styled`, right?

This seems strange to me! Are there any error logs in the Browser console when you reset the properties? If the select is using `wire:model` and you reset the property, then the selected value should be automatically removed as in my tests.

I'm not sure about the `x-on:change` event, but `wire:change` is accepted without problems in the select styled component.

**Comment by ldommer** [comment]
Yes, I'm using the styled select component. There aren't any bugs in the browser console.

![Screenshot 2024-03-15 at 22-14-10 Artikel - Übersicht Thomas Philipps Marktportal für Pflanzenversteigerungen](https://github.com/tallstackui/tallstackui/assets/65616313/1d9b5547-3da3-4591-bd8c-3f5ef9078617)

This is the way my UI looks after I've reset the Livewire props. The state of my component is reset, the table has rerendered without the filters applied. But the select fields still have their options selected in the UI.

**Reply by devajmeireles** [reply]
Could you share the source code with me, please? I'll need to try to reproduce this issue locally.

**Reply by ldommer** [reply]
Component:

```php
<?php

namespace App\Livewire\Article;

use App\Enums\OrderQuantity;
use App\Models\Article;
use App\Models\ArticleCategory;
use Illuminate\Contracts\View\View;
use Illuminate\Database\Eloquent\Builder;
use Livewire\WithPagination;

class Index
{
    use WithPagination;

    public string $search = '';
    public array $articleCategories;
    public string $articleCategoryId = '';
    public array $orderQuantities;
    public string $orderQuantity = '';
    public array $resetExcept = ['articleCategories', 'orderQuantities'];

    public function mount(): void
    {
        $this->articleCategories = ArticleCategory::orderBy('name')
            ->get()
            ->map(fn (ArticleCategory $articleCategory) => [
                'value' => $articleCategory->id,
                'label' => $articleCategory->name,
            ])
            ->toArray();
        $this->orderQuantities = OrderQuantity::selectOptions();
    }

    public function render(): View
    {
        $articles = Article::with('articleCategory')
            ->withCount('offers')
            ->when(!empty($this->search), function (Builder $q) {
                $q->where('id', $this->search)
                    ->orWhereHas('articleCategory', function (Builder $q) {
                        $q->where('name', 'like', '%' . $this->search . '%');
                    })
                    ->orWhere('name', 'like', '%' . $this->search . '%')
                    ->orWhere('description', 'like', '%' . $this->search . '%');
            })
            ->when(!empty($this->articleCategoryId), function (Builder $q) {
                $q->where('article_category_id', $this->articleCategoryId);
            })
            ->when(!empty($this->orderQuantity), function (Builder $q) {
                $q->where('order_quantities', 'like', '%' . $this->orderQuantity . '%');
            })
            ->oldest('name')
            ->oldest('id')
            ->paginate(10);

        return view('livewire.article.index')
            ->with(compact('articles'));
    }

    public function updating(): void
    {
        $this->resetPage();
    }

    public function resetSearchForm(): void
    {
        $this->resetExcept($this->resetExcept);
    }
}
```

Template:
```html
@php
    use App\Models\Article;
    use App\Models\ArticleCategory;
@endphp

<div class="space-y-6">
    <x-table.search-form search-title="{{ trans('articles.search') }}" :model-class="Article::class"
        create-route="{{ route('backend.articles.create') }}" cols="3">
        <x-form.select id="articleCategoryId" :options="$articleCategories" :required="false"
            placeholder="{{ trans_choice('article_categories.title', 1) }}" wire:model.live="articleCategoryId" />
        <x-form.select id="orderQuantity" :options="$orderQuantities" :required="false"
            placeholder="{{ trans_choice('order_quantities.title', 1) }}" wire:model.live="orderQuantity" />
    </x-table.search-form>
    <div class="table-responsive">
        <table>
            <thead>
                <tr>
                    <th scope="col">{{ trans('app.id') }}</th>
                    <th scope="col">{{ trans_choice('article_categories.title', 1) }}</th>
                    <th scope="col">{{ trans('app.name') }}</th>
                    <th scope="col">{{ trans('articles.size') }}</th>
                    <th scope="col">{{ trans('articles.weight') }}</th>
                    <th scope="col">{{ trans_choice('order_quantities.title', 2) }}</th>
                    <th scope="col">{{ trans_choice('offers.title', 2) }}</th>
                    <th scope="col">{{ trans('app.actions') }}</th>
                </tr>
            </thead>
            <tbody class="bg-white">
                @forelse($articles as $article)
                    <tr class="even:bg-slate-50" wire:key="{{ $article->id }}">
                        <td>{{ $article->id }}</td>
                        <td>
                            @can('view', $article->articleCategory)
                                <x-generic.link
                                    href="{{ route('backend.article_categories.show', $article->articleCategory) }}"
                                    :title="$article->articleCategory->name" />
                            @else
                                {{ $article->articleCategory->name }}
                            @endcan
                        </td>
                        <td>{{ $article->name }}</td>
                        <td>{{ $article->size() }}</td>
                        <td>{{ $article->weight() }}</td>
                        <td>
                            @foreach ($article->orderQuantities() as $orderQuantity)
                                <x-generic.badge :color="$orderQuantity->color()" :title="$orderQuantity->name()" />
                            @endforeach
                        </td>
                        <td>{{ $article->offers_count }}</td>
                        <td>
                            <x-table.action-buttons :model="$article" :show-route="route('backend.articles.show', $article)" :edit-route="route('backend.articles.edit', $article)" />
                        </td>
                    </tr>
                @empty
                    <x-table.empty-row :colspan="8" />
                @endforelse
            </tbody>
        </table>
    </div>
    {{ $articles->links() }}
</div>
```

Blade Select Component:
```html
@props(['title' => null, 'id', 'options', 'required' => true, 'select' => 'label:label|value:value'])

<div class="w-full max-w-full">
    @if (!empty($title))
        <x-form.label :$required />
    @endif
    <x-ts-select.styled :$id invalidate :$options :$required select="{{ $select }}"
        {{ $attributes->merge(['class' => 'w-full']) }} />
    <x-form.errors :$id />
</div>

**Comment by ldommer** [comment]
@devajmeireles

Today I upgraded to the latest version, because I wanted to use one of your new components. The issue is gone. Before I was running an older version from mid / end of january. Hopefully I didn't cause too much trouble...

**Comment by devajmeireles** [comment]
> @devajmeireles
> 
> Today I upgraded to the latest version, because I wanted to use one of your new components. The issue is gone. Before I was running an older version from mid / end of january. Hopefully I didn't cause too much trouble...

Happy to know the issue is gone. Feel free to open new issues whenever necessary.
